### ISSUE TITLE
CI: sigprocmask-32 fails

### ISSUE THREAD

**Issue by bernhardu** [rr-debugger/rr#3441]
I noticed that the buildkite log shows the test `sigprocmask-32` failing (e.g [here](https://buildkite.com/julialang/rr/builds/965#01866d70-f1b0-4ded-8995-020e3f550a16/4355-6705)).
```
[FATAL src/ReplaySession.cc:678:check_pending_sig()]
 (task 117252 (rec:116819) at time 172)
 -> Assertion `false' failed to hold. Replaying `SIGNAL: SIGUSR1(async)': expecting tracee signal or trap, but instead at `write' (ticks: 33608609)
```

And I was able to reproduce this failure inside a Ubuntu 18.08 VM.

A bisect led to comit 9fea9c77f5b.

And running with following RR_LOG below lines can be seen which seem related, especially the `compare regs`:
```
$ RR_LOG=ExtraRegisters,ReplaySession bin/rr replay -a sigprocmask_32-9
...
[ReplaySession] [event 166] 1188: replaying SYSCALL: sigprocmask; state EXITING_SYSCALL
[ReplaySession]     (syscllbufsz:0, abrtcmt:0, locked:0)
[ReplaySession] [event 167] 1188: replaying SIGNAL: SIGUSR1(async); state  (none)
[ReplaySession]     (syscllbufsz:0, abrtcmt:0, locked:0)
[ReplaySession] advancing 0 ticks to reach 33608510/0x70000002
[INFO compare_regs()] ymm0 200 != 0 (rep vs. rec)
[INFO compare_regs()] ymm1 200 != 0 (rep vs. rec)
[INFO compare_regs()] ymm2 ffffffffffff7fff != 0 (rep vs. rec)
[ReplaySession]   not same execution point: extra regs differ (@0x70000002)
[ReplaySession]     (fast-forwarding over target $ip)
[ReplaySession]   not same execution point: regs differ (@0x70000002)
[ReplaySession]   not same execution point: regs differ (@0x70000002)
[ReplaySession]     (SIGTRAP; stepi'd target $ip)
[ReplaySession]     breaking on target $ip
[FATAL src/ReplaySession.cc:678:check_pending_sig()] 
 (task 8959 (rec:1188) at time 167)
 -> Assertion `false' failed to hold. Replaying `SIGNAL: SIGUSR1(async)': expecting tracee signal or trap, but instead at `write' (ticks: 33608885)
...
```

I tested updating the kernel `5.4.0-135-generic #152~18.04.2-Ubuntu` to some
mainline kernel versions from [here](https://kernel.ubuntu.com/~kernel-ppa/mainline/?C=N;O=D) and found
it failing until 5.8.18, and starting with version 5.9 and up the error is gone.

Recordings taken with older kernel work with the newer kernels, so I assume
the recording is good.

Just wanted to provide the information so far.
Is this a known issue, should there be done more investigation - I would
like to avoid bisecting the kernel if there is no benefit of it ...

**Comment by khuey** [comment]
Thanks, that's helpful.

@Keno, any chance we could get a newer kernel on CI?

**Comment by Keno** [comment]
> @Keno, any chance we could get a newer kernel on CI?

That's a question for @staticfloat.

**Comment by bernhardu** [comment]
@khuey: Would something like a (to be invented) `skip_if_kernel_below(5.9)` be a good option?

**Comment by khuey** [comment]
Yeah that would be fine too.

**Comment by staticfloat** [comment]
We have the CI machines slated for an OS update to Ubuntu 22.04 which should get us a kernel version of v5.15.  I'm not sure when the OS update will happen (it's being done by the managers of the machine at MIT, not by us) but I will let you guys know when it's done.

**Comment by khuey** [comment]
Do you happen have a general idea on the time frame (weeks, months, etc?)

**Comment by bernhardu** [comment]
I did restart my 18.04 VM today and could no longer reproduce the issue.
The difference was today I used the build notes from rr wiki,
but yesterday I used my slightly different personal notes.
I compared both and found builds
with `cmake -DCMAKE_BUILD_TYPE=Release ../rr` do not show the issue, but
with `cmake -DCMAKE_BUILD_TYPE=RelWithDebInfo ../rr` they do (for me).
Unfortunately I could not find which type the CI uses.

**Comment by bernhardu** [comment]
Some further comparing revealed that the difference is the optimization level,
Release defaults to `-O3` but RelWithDebInfo defaults to `-O2`.
The CI uses build type [Debug](https://buildkite.com/julialang/rr/builds/969#0186808a-f60c-4465-9392-999b7a46486d/398-420), but for this `-O2` is added manually to PRELOAD_COMPILE_FLAGS.

I could further isolate it to `librrpreload_32.so` and there the fault vanishes
if at least `syscallbuf.c` is compiled with `-O3`.
A little more trial and error showed that it it is sufficient to just optimize the function `init_process` with `-O3` like below.
```
#pragma GCC push_options
#pragma GCC optimize ("-O3")
...
#pragma GCC pop_options
```

But I guess changing just optimization will still not count as solution yet?

**Comment by rocallahan** [comment]
I suspect the optimization level makes syscallbuf.c use XMM registers in a different way. Care to dump the disassembly of `init_thread` using `objdump -d` with -O2 vs -O3?

**Comment by rocallahan** [comment]
https://github.com/rr-debugger/rr/commit/67b503eb4656deb5e52c64b0d6441f2e32ec7427 might help with debugging. Would be useful to see what the compare_regs log says now.

**Comment by bernhardu** [comment]
If it is still of interest:
```diff
--- /tmp/init_process_-O3.txt   2023-02-24 10:53:36.773274600 +0100
+++ /tmp/init_process_-O2.txt   2023-02-24 10:53:36.749274700 +0100
@@ -1,5 +1,5 @@
-/usr/bin/cc -DFANOTIFY_H=1 -DMQUEUE_H=1 -DPROC_SERVICE_H=1 -DRR_VERSION=\"5.6.0\" -Drrpreload_32_EXPORTS -I/home/benutzer/rr/include -I/home/benutzer/rr/third-party/proc-service -I/home/benutzer/rr/third-party/brotli/include -I/home/benutzer/obj-works  -D__USE_LARGEFILE64 -pthread -msse2 -D__MMX__ -D__SSE__ -D__SSE2__ -Wstrict-prototypes -std=gnu11  -pthread  -O3 -DNDEBUG -fPIC   -m32 -Wall -Wextra -UDEBUG -DNDEBUG -fno-stack-protector -g3 -U_FORTIFY_SOURCE -o CMakeFiles/rrpreload_32.dir/32/preload/syscallbuf.c-O3.o   -c /home/benutzer/obj-works/32/preload/syscallbuf.c
-objdump -d CMakeFiles/rrpreload_32.dir/32/preload/syscallbuf.c-O3.o
+/usr/bin/cc -DFANOTIFY_H=1 -DMQUEUE_H=1 -DPROC_SERVICE_H=1 -DRR_VERSION=\"5.6.0\" -Drrpreload_32_EXPORTS -I/home/benutzer/rr/include -I/home/benutzer/rr/third-party/proc-service -I/home/benutzer/rr/third-party/brotli/include -I/home/benutzer/obj-works  -D__USE_LARGEFILE64 -pthread -msse2 -D__MMX__ -D__SSE__ -D__SSE2__ -Wstrict-prototypes -std=gnu11  -pthread  -O2 -DNDEBUG -fPIC   -m32 -Wall -Wextra -UDEBUG -DNDEBUG -fno-stack-protector -g3 -U_FORTIFY_SOURCE -o CMakeFiles/rrpreload_32.dir/32/preload/syscallbuf.c-O2.o   -c /home/benutzer/obj-works/32/preload/syscallbuf.c
+objdump -d CMakeFiles/rrpreload_32.dir/32/preload/syscallbuf.c-O2.o
 
 Disassembly of section .text.startup:
 
@@ -12,43 +12,43 @@ Disassembly of section .text.startup:
    e8 fc ff ff ff          call   b <init_process+0xb>
    81 c3 02 00 00 00       add    $0x2,%ebx
    83 ec 70                sub    $0x70,%esp
-   89 e6                   mov    %esp,%esi
-   c7 44 24 2c 00 00 00    movl   $0x0,0x2c(%esp)
+   8d 74 24 0c             lea    0xc(%esp),%esi
+   c7 44 24 38 00 00 00    movl   $0x0,0x38(%esp)
    00 
    89 f7                   mov    %esi,%edi
    f3 ab                   rep stos %eax,%es:(%edi)
    8d 83 00 00 00 00       lea    0x0(%ebx),%eax
    bf ff ff ff ff          mov    $0xffffffff,%edi
-   c7 44 24 01 05 3d 01    movl   $0xf0013d05,0x1(%esp)
+   c7 44 24 0d 05 3d 01    movl   $0xf0013d05,0xd(%esp)
    f0 
-   66 89 7c 24 05          mov    %di,0x5(%esp)
-   c7 44 24 14 00 00 00    movl   $0x0,0x14(%esp)
+   66 89 7c 24 11          mov    %di,0x11(%esp)
+   c7 44 24 20 00 00 00    movl   $0x0,0x20(%esp)
    00 
-   89 44 24 10             mov    %eax,0x10(%esp)
+   89 44 24 1c             mov    %eax,0x1c(%esp)
    8d 83 00 00 00 00       lea    0x0(%ebx),%eax
-   c7 44 24 18 01 03 90    movl   $0x90900301,0x18(%esp)
+   c7 44 24 24 01 03 90    movl   $0x90900301,0x24(%esp)
    90 
-   c6 44 24 1c 90          movb   $0x90,0x1c(%esp)
-   89 44 24 28             mov    %eax,0x28(%esp)
+   c6 44 24 28 90          movb   $0x90,0x28(%esp)
+   89 44 24 34             mov    %eax,0x34(%esp)
    8b 83 08 00 00 00       mov    0x8(%ebx),%eax
    85 c0                   test   %eax,%eax
-   75 3d                   jne    a8 <init_process+0xa8>
+   75 3d                   jne    aa <init_process+0xaa>
    83 bb 00 00 00 00 00    cmpl   $0x0,0x0(%ebx)
-   74 14                   je     88 <init_process+0x88>
+   74 14                   je     8a <init_process+0x8a>
    83 ec 0c                sub    $0xc,%esp
    6a 21                   push   $0x21
-   e8 fc ff ff ff          call   7a <init_process+0x7a>
+   e8 fc ff ff ff          call   7c <init_process+0x7c>
    83 c4 10                add    $0x10,%esp
    3d 00 d0 ff 6f          cmp    $0x6fffd000,%eax
-   74 27                   je     af <init_process+0xaf>
+   74 27                   je     b1 <init_process+0xb1>
    51                      push   %ecx
    6a 01                   push   $0x1
    6a 01                   push   $0x1
    68 00 00 00 70          push   $0x70000000
-   e8 fc ff ff ff          call   93 <init_process+0x93>
+   e8 fc ff ff ff          call   95 <init_process+0x95>
    83 c4 10                add    $0x10,%esp
    85 c0                   test   %eax,%eax
-   74 11                   je     af <init_process+0xaf>
+   74 11                   je     b1 <init_process+0xb1>
    c7 83 0c 00 00 00 00    movl   $0x0,0xc(%ebx)
    00 00 00 
    83 c4 70                add    $0x70,%esp
@@ -59,74 +59,69 @@ Disassembly of section .text.startup:
    8d 83 00 00 00 00       lea    0x0(%ebx),%eax
    83 ec 0c                sub    $0xc,%esp
    50                      push   %eax
-   e8 fc ff ff ff          call   ba <init_process+0xba>
+   e8 fc ff ff ff          call   bc <init_process+0xbc>
    31 d2                   xor    %edx,%edx
    83 c4 10                add    $0x10,%esp
    85 c0                   test   %eax,%eax
    0f 95 c2                setne  %dl
    89 93 0c 00 00 00       mov    %edx,0xc(%ebx)
-   74 d8                   je     a8 <init_process+0xa8>
-   66 0f 6e 83 00 00 00    movd   0x0(%ebx),%xmm0
-   00 
+   74 d8                   je     aa <init_process+0xaa>
    8b 83 00 00 00 00       mov    0x0(%ebx),%eax
-   66 0f 6e 9b 00 00 00    movd   0x0(%ebx),%xmm3
-   00 
-   c7 44 24 30 01 00 00    movl   $0x1,0x30(%esp)
-   00 
-   66 0f 6e 8b 00 00 00    movd   0x0(%ebx),%xmm1
-   00 
-   c7 44 24 34 02 00 00    movl   $0x2,0x34(%esp)
+   c7 44 24 3c 01 00 00    movl   $0x1,0x3c(%esp)
    00 
-   66 0f 6e 93 00 00 00    movd   0x0(%ebx),%xmm2
+   c7 44 24 40 02 00 00    movl   $0x2,0x40(%esp)
    00 
-   66 0f 62 d8             punpckldq %xmm0,%xmm3
-   89 44 24 50             mov    %eax,0x50(%esp)
-   8d 83 00 00 00 00       lea    0x0(%ebx),%eax
-   89 74 24 38             mov    %esi,0x38(%esp)
-   66 0f 62 d1             punpckldq %xmm1,%xmm2
-   c7 44 24 5c ff ff ff    movl   $0xffffffff,0x5c(%esp)
+   89 74 24 44             mov    %esi,0x44(%esp)
+   c7 44 24 68 ff ff ff    movl   $0xffffffff,0x68(%esp)
    ff 
+   c6 83 d8 61 00 00 01    movb   $0x1,0x61d8(%ebx)
    89 44 24 54             mov    %eax,0x54(%esp)
-   66 0f 6f c3             movdqa %xmm3,%xmm0
    8b 83 00 00 00 00       mov    0x0(%ebx),%eax
-   c6 83 d8 61 00 00 01    movb   $0x1,0x61d8(%ebx)
-   66 0f 6c c2             punpcklqdq %xmm2,%xmm0
    89 44 24 58             mov    %eax,0x58(%esp)
+   8b 83 00 00 00 00       mov    0x0(%ebx),%eax
+   89 44 24 4c             mov    %eax,0x4c(%esp)
+   8b 83 00 00 00 00       mov    0x0(%ebx),%eax
+   89 44 24 50             mov    %eax,0x50(%esp)
+   8b 83 00 00 00 00       mov    0x0(%ebx),%eax
+   89 44 24 5c             mov    %eax,0x5c(%esp)
    8d 83 00 00 00 00       lea    0x0(%ebx),%eax
-   0f 29 44 24 40          movaps %xmm0,0x40(%esp)
    89 44 24 60             mov    %eax,0x60(%esp)
+   8b 83 00 00 00 00       mov    0x0(%ebx),%eax
+   89 44 24 64             mov    %eax,0x64(%esp)
+   8d 83 00 00 00 00       lea    0x0(%ebx),%eax
+   89 44 24 6c             mov    %eax,0x6c(%esp)
    52                      push   %edx
    6a 00                   push   $0x0
    6a 00                   push   $0x0
    6a 00                   push   $0x0
    6a 00                   push   $0x0
    6a 00                   push   $0x0
-   8d 44 24 48             lea    0x48(%esp),%eax
+   8d 44 24 54             lea    0x54(%esp),%eax
    50                      push   %eax
    68 e8 03 00 00          push   $0x3e8
-   e8 fc ff ff ff          call   166 <init_process+0x166>
+   e8 fc ff ff ff          call   15b <init_process+0x15b>
    83 c4 20                add    $0x20,%esp
    85 c0                   test   %eax,%eax
-   75 0f                   jne    180 <init_process+0x180>
+   75 0f                   jne    175 <init_process+0x175>
    c7 83 08 00 00 00 01    movl   $0x1,0x8(%ebx)
    00 00 00 
-   e9 28 ff ff ff          jmp    a8 <init_process+0xa8>
+   e9 35 ff ff ff          jmp    aa <init_process+0xaa>
    50                      push   %eax
    6a 01                   push   $0x1
    6a 01                   push   $0x1
    68 00 10 00 70          push   $0x70001000
-   e8 fc ff ff ff          call   18b <init_process+0x18b>
+   e8 fc ff ff ff          call   180 <init_process+0x180>
    83 c4 10                add    $0x10,%esp
    85 c0                   test   %eax,%eax
-   0f 85 04 ff ff ff       jne    9e <init_process+0x9e>
+   0f 85 11 ff ff ff       jne    a0 <init_process+0xa0>
    8d 83 00 00 00 00       lea    0x0(%ebx),%eax
    89 c1                   mov    %eax,%ecx
    83 c1 01                add    $0x1,%ecx
    89 ca                   mov    %ecx,%edx
    29 c2                   sub    %eax,%edx
    80 39 00                cmpb   $0x0,(%ecx)
-   75 f4                   jne    1a2 <init_process+0x1a2>
-   e8 fc ff ff ff          call   1af <init_process+0x1af>
-   e8 cc 02 00 00          call   484 <start_commit_buffered_syscall.part.6+0x24>
+   75 f4                   jne    197 <init_process+0x197>
+   e8 fc ff ff ff          call   1a4 <init_process+0x1a4>
+   e8 1c 03 00 00          call   4c9 <commit_raw_syscall+0xe9>
 
-Disassembly of section .text.__x86.get_pc_thunk.ax:
+Disassembly of section .text.__x86.get_pc_thunk.dx:
```

**Comment by rocallahan** [comment]
-O3 uses XMM while -O2 does not, so that's a good hint. I don't see how init_process behaviour affects later execution though. Can you attach the `rr dump -b -m -p` of the recording that fails to replay?

**Comment by bernhardu** [comment]
I am no longer able to reproduce it after 67b503eb4.
Also the [CI run](https://buildkite.com/julialang/rr/builds/972) for this commit did no longer show this failure.

Following file contains the dump, a replay with RR_LOG=all, and the record directory after pack,
from a build at git 0c0f3f737c, from the Ubuntu 18.04 VM running below an Ryzen 1700.
[sigprocmask_32-50.tar.gz](https://github.com/rr-debugger/rr/files/10823247/sigprocmask_32-50.tar.gz)

Because 67b503eb4 looks like it had already a positive effect, maybe consider this fixed? (if the next CI runs succeed)

**Comment by rocallahan** [comment]
You could try adding this patch:
```
diff --git a/src/Event.cc b/src/Event.cc
index a855533f..87c2657a 100644
--- a/src/Event.cc
+++ b/src/Event.cc
@@ -94,6 +94,7 @@ bool Event::record_regs() const {
 }
 
 bool Event::record_extra_regs() const {
+  return record_regs();
   switch (type()) {
     case EV_SYSCALL: {
       const SyscallEvent& sys_ev = Syscall();
```
and rerecord/replay ... should narrow down the problem.

I suspect we're going to find a kernel bug that was fixed in 5.9. It might not be worth the work to find it.

**Comment by bernhardu** [comment]
> I suspect we're going to find a kernel bug that was fixed in 5.9. It might not be worth the work to find it.

To be on the safe side I went through the bisection (inside a 20.04 VM):

possible first fixed commit: [0557d64d983e3dedead2d4b4e8abc49620d5f5d2](https://github.com/torvalds/linux/commit/0557d64d983e3dedead2d4b4e8abc49620d5f5d2) x86: switch to ->regset_get()
possible first fixed commit: [7717cb9bdd0421faa432a4e0d499fdba6e2394c8](https://github.com/torvalds/linux/commit/7717cb9bdd0421faa432a4e0d499fdba6e2394c8) regset: new method and helpers for it
possible first fixed commit: [dc12d7968f9c9540494deb1285854b18ca4465ec](https://github.com/torvalds/linux/commit/dc12d7968f9c9540494deb1285854b18ca4465ec) copy_regset_to_user(): do all copyout at once.

Those three commits do compile and boot separately, but rr fails to record with 7717cb9bdd04 and dc12d7968f9c. 

(If someone needs to do another bisect between v5.8 and v5.9 then those unrelated regressions are to be applied where needed: [cc17b22559d9](https://github.com/torvalds/linux/commit/cc17b22559d9b9c8b7540810df172f3d7af901ce) kernel does not boot,  [d88d59b64ca3](https://github.com/torvalds/linux/commit/d88d59b64ca35abae208e2781fdb45e69cbed56c) rr cannot record, [16b89f695313](https://github.com/torvalds/linux/commit/16b89f695313d91c99bdaf6898f28a51d0af1b17) rr cannot record.)


> You could try adding this patch:
> and rerecord/replay ... should narrow down the problem.

If I see it right this adds the XMM register to the trace dump. But I found all registers just zeros all the time without 67b503eb465.
With 67b503eb465 the test succeeds.

**Comment by bernhardu** [comment]
The [CI builds page](https://buildkite.com/julialang/rr/builds) shows now 9 runs since 972 where the sigprocmask-32 test succeeded. Therefore I hope it is safe to close this issue.
